clang/lib/Parse/ParseTentative.cpp

```diff
@@ -14,6 +14,8 @@
 #include "clang/Basic/DiagnosticParse.h"
 #include "clang/Parse/Parser.h"
 #include "clang/Sema/ParsedTemplate.h"
+#include "clang/Tooling/Inclusions/StandardLibrary.h"
```

Comment:
This seems like a layering violation, though not a circular dependency. Parser currently does not link against tooling. In your patch, you link tooling into Sema which Parser then picks up transitively. Tooling does not currently link in Parser or Sema, hence no circular dependency.

I kind of think the standard library stuff needs to be lifted into AST. It does seem to rely on `Decl.h` internally, so it can't be lifted into Basic.

CC @erichkeane for other opinions

Reply:
Sure, but we want to rewrite anyway. The implementation (StandardLibrary.cpp) seems fairly inefficient. We should try to TableGen a trie and store version information in there, have some support for typo correction, etc